Frameworks/CoreText/CTFramesetter.mm

```diff
@@ -27,15 +27,48 @@ @implementation _CTFramesetter : NSObject
 @end
 
 static _CTFrame* __CreateFrame(_CTFramesetter* framesetter, CGRect frameSize, CFRange range) {
-    RETURN_NULL_IF(!framesetter);
+    RETURN_NULL_IF(framesetter == nil);
 
     // Call _DWriteWrapper to get _CTLine object list that makes up this frame
     _CTTypesetter* typesetter = static_cast<_CTTypesetter*>(framesetter->_typesetter);
     if (range.length == 0L) {
         range.length = typesetter->_characters.size();
     }
 
-    return [_DWriteGetFrame(static_cast<CFAttributedStringRef>(typesetter->_attributedString.get()), range, frameSize) retain];
+    _CTFrame* ret = _DWriteGetFrame(static_cast<CFAttributedStringRef>(typesetter->_attributedString.get()), range, frameSize);
+
+    // Trying to access attributes without any text with throw an error
+    if (range.length != 0L) {
+        CTParagraphStyleRef settings =
+            static_cast<CTParagraphStyleRef>([typesetter->_attributedString attribute:static_cast<NSString*>(kCTParagraphStyleAttributeName)
+                                                                              atIndex:0
+                                                                       effectiveRange:nullptr]);
+        if (settings) {
+            // DWrite only gives manual control of lineheight when it is constant through a frame
+            // We need to shift each line by the difference in lineheight manually
+            CGFloat lineHeightMultiple = 0.0f;
+            if (CTParagraphStyleGetValueForSpecifier(settings,
+                                                     kCTParagraphStyleSpecifierLineHeightMultiple,
+                                                     sizeof(lineHeightMultiple),
+                                                     &lineHeightMultiple) &&
+                lineHeightMultiple > 0) {
+                // The actual ratio we need to change the line height by is multiple - 1
+                lineHeightMultiple -= 1.0f;
+                CGFloat totalShifted = 0.0f;
+                CGFloat lastOriginY = frameSize.origin.y;
+                for (size_t i = 0; i < ret->_lineOrigins.size(); ++i) {
+                    totalShifted += lineHeightMultiple * (ret->_lineOrigins[i].y - lastOriginY);
+                    lastOriginY = ret->_lineOrigins[i].y;
+                    ret->_lineOrigins[i].y += totalShifted;
+                }
+
+                // Adjust framesize to account for changes in lineheights
+                ret->_frameRect.size.height += totalShifted;
+            }
+        }
+    }
```

Comment:
Although you don't "prefer" early returns. This code would be so much simpler and much simpler to read if you used early returns.
We have if layers that aren't the greatest looking for code readability.